File: 0270. Closest Binary Search Tree Value/Solution.cpp
#include <cmath>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode()
        : val(0)
        , left(nullptr)
        , right(nullptr)
    {
    }
    TreeNode(int x)
        : val(x)
        , left(nullptr)
        , right(nullptr)
    {
    }
    TreeNode(int x, TreeNode* left, TreeNode* right)
        : val(x)
        , left(left)
        , right(right)
    {
    }
};

class Solution {
public:
    int closestValue(TreeNode* root, double target)
    {
        int res = root->val;
        while (root) {
            if (abs(root->val - target) < abs(res - target) || (abs(root->val - target) == abs(res - target) && root->val < res))
                res = root->val;
            root = root->val > target ? root->left : root->right;
        }
        return res;
    }
};

File: 0270. Closest Binary Search Tree Value/Solution.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def closestValue(self, root: Optional[TreeNode], target: float) -> int:
        res = root.val
        while root:
            if abs(root.val - target) < abs(res - target) or (
                abs(root.val - target) == abs(res - target) and root.val < res
            ):
                res = root.val
            if root.val > target:
                root = root.left
            else:
                root = root.right
        return res
